### PR TITLE
pkg/provides/vault/many: allow specifying retry logic

### PR DESCRIPTION
**Pull request by anonymouse64**
Allow a SecretClient to be created that retries HTTP requests that are not 2XX. Note that this behavior is optional, and needs to be opt-in by the calling client.

This enables clients of go-mod-secrets to easily block waiting for their secrets to be available.

Also include a proposed implementation of "infinite" blocking by setting the AdditionalRetryAttempts to a negative number, but error out for now, so as to not complicate testing.

Finally, update the tests to check the number of calls to Do() that are made when getting secrets under various situations.

To test this change IRL, see https://github.com/edgexfoundry/docker-edgex-mongo/pull/59 and the testing instructions there.

### CODE REVIEW

**Comment by anonymouse64** [comment]
CC @tingyuz @bnevis-i @jim-wang-intel @mkbhanda

File: pkg/providers/vault/client.go [hunk 1]
```diff
@@ -90,13 +91,63 @@ func (c Client) getAllKeys(path string) (map[string]string, error) {
...
-		return nil, pkg.NewErrSecretStore(fmt.Sprintf("Received a '%d' response from the secret store", resp.StatusCode))
+		if resp.StatusCode < 200 || resp.StatusCode > 299 {
+			return nil, pkg.NewErrSecretStore(fmt.Sprintf("Received a '%d' response from the secret store", resp.StatusCode))
+		}
+
+	case addRetryAttempts > 0:
+		// do some retries
+		// note the limit is 1 + additional retry attempts, cause we always need
+		// to do 1
+		for tryNum := 0; tryNum < 1+addRetryAttempts; tryNum++ {
```
Comment (AnthonyMBonafide) [inline pkg/providers/vault/client.go]: I think the retry logic should be moved to the `GetSecrets` method:

https://github.com/edgexfoundry/go-mod-secrets/blob/65a79df9754eb7cfd29c29606b68a90d53ede380/pkg/providers/vault/client.go#L49

During testing I ran into a situation where there were some secrets in Vault, but not all of them. This resulted in the `getAllKeys` successfully returning and the `GetSecrets` method would then return an error, bypassing the retry logic. By moving it to `GetSecrets` we get the retry logic to work with any failing operation while retrieving keys.

File: pkg/providers/vault/client.go [hunk 2]
```diff
@@ -90,13 +91,63 @@ func (c Client) getAllKeys(path string) (map[string]string, error) {
...
-	resp, err := c.HttpCaller.Do(req)
-	if err != nil {
-		return nil, err
-	}
+	var resp *http.Response
+	addRetryAttempts := c.HttpConfig.AdditionalRetryAttempts
+	switch {
+	case addRetryAttempts < 0:
+		return nil, pkg.NewErrSecretStore(fmt.Sprintf("invalid retry attempts setting %d", addRetryAttempts))
+		// the number of retries is infinite
```
Comment (AnthonyMBonafide) [inline pkg/providers/vault/client.go]: Can we remove these comments?

**Comment by AnthonyMBonafide** [comment]
@tingyuz , I submitted my review too late. If these changes make sense I can do a follow-up PR to address these comments. Any thoughts?

File: pkg/providers/vault/client.go (hunk 2 above)
Comment (anonymouse64) [reply pkg/providers/vault/client.go]: sure

File: pkg/providers/vault/client.go (hunk 1 above)
Comment (anonymouse64) [reply pkg/providers/vault/client.go]: okay, I can move this to GetSecrets, when I initially worked on this I wanted to be able to only retry on http errors and not on other catastrophic errors, I will submit a PR a bit later today doing this

File: pkg/providers/vault/config.go [hunk 3]
```diff
@@ -17,18 +17,21 @@ package vault
...
+	Host                    string
+	Port                    int
+	Path                    string
+	Protocol                string
+	Namespace               string
+	RootCaCertPath          string
+	ServerName              string
+	Authentication          AuthenticationInfo
+	AdditionalRetryAttempts int
+	RetryWaitPeriod         time.Duration
```
Comment (AnthonyMBonafide) [inline pkg/providers/vault/config.go]: During my testing while integrating this with `edgex-go` this configuration was being treated as a `nanosecond` value. For example, I provided :
```
...
AdditionalRetryAttempts = 5
RetryWaitPeriod = 10
...
```

Is this intended or is my configuration values incorrect?

Comment (anonymouse64) [reply pkg/providers/vault/config.go]: you need to pass a time.Duration including the units, so if you wanted 10 nanoseconds, the "right" way to do this would be:

```
RetryWaitPeriod = 10*time.Nanosecond
```
or if you wanted 10 seconds, then you would do

```
RetryWaitPeriod = 10*time.Second
```

See https://play.golang.org/p/Uo7UP9SwrOj for an example

it is kinda unfortunate that time.Duration is really just type-cast as an int and internally stored as nanoseconds (with the time.Second, etc. constants just being defined as the number of nanoseconds per that unit, so time.Second becomes 1000000000), because it means specifying just raw numbers is accepted blindly as nanoseconds which is almost always what folks _don't_ mean. OTOH it does make arithmetic with time.Duration quite easy :man_shrugging: 

TLDR, yes you're holding it wrong :smile:

Comment (AnthonyMBonafide) [reply pkg/providers/vault/config.go]: How would we specify that in a configuration TOML? `edgex-go` basically passed the `SecretConfig` values though to `go-mod-secrets` without any additional processing.

Comment (anonymouse64) [reply pkg/providers/vault/config.go]: Oh I see... you want to specify this through configuration.toml.

What I'd recommend there instead is to just have the RetryWaitPeriod in configuration.toml be a time string, parse it into a time.Duration with [`ParseDuration`](https://golang.org/pkg/time/#ParseDuration)

We could accept a string here and do the parsing inside go-mod-secrets but I don't like doing that because having a time.Duration exposed from a Go interface perspective is definitely the right:tm: thing to do from a Go programming perspective, but if that's what you'd prefer I can refactor to do that as well I suppose.

**Comment by anonymouse64** [comment]
@AnthonyMBonafide PR with your feedback is up here: https://github.com/edgexfoundry/go-mod-secrets/pull/37
